src/libAtomVM/nifs.c

```diff
@@ -760,14 +767,38 @@ static term nif_erlang_register_2(Context *ctx, int argc, term argv[])
     VALIDATE_VALUE(pid_or_port_term, term_is_pid);
 
     int atom_index = term_to_atom_index(reg_name_term);
-    int pid = term_to_local_process_id(pid_or_port_term);
+    int32_t pid = term_to_local_process_id(pid_or_port_term);
 
-    // TODO: pid must be existing, not already registered.
-    // TODO: reg_name_term must not be the atom undefined and not already registered.
+    // pid must be existing, not already registered.
+    if (UNLIKELY(globalcontext_get_process(ctx->global, pid) == NULL)){
+        RAISE_ERROR(BADARG_ATOM);
+    } else if (UNLIKELY(globalcontext_get_registered_process(ctx->global, atom_index) != 0)) {
+        RAISE_ERROR(BADARG_ATOM);
+    // reg_name_term must not be the atom undefined.
+    } else if (UNLIKELY(atom_index == UNDEFINED_ATOM_INDEX)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
 
     globalcontext_register_process(ctx->global, atom_index, pid);
 
-    return term_nil();
+    return TRUE_ATOM;
+}
+
+static term nif_erlang_unregister_1(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+
+    term reg_name_term = argv[0];
+    VALIDATE_VALUE(reg_name_term, term_is_atom);
+
+    int atom_index = term_to_atom_index(reg_name_term);
+
+    bool ret = globalcontext_unregister_process(ctx->global, atom_index);
+    if (ret) {
+        return TRUE_ATOM;
+    }
```

Comment:
I know it is correct, but it looks weird, we should rather early return on error, and handle `return TRUE_ATOM` as the normal path.

Reply:
Also `if (UNLIKELY(!ret))`.